lib/utils/interval.go

```diff
@@ -0,0 +1,131 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+// Interval functions similarly to time.Ticker, with the added benefit of being
+// able to specify a custom duration for the first "tick", and an optional
+// per-tick jitter.  When attempting to stagger periodic operations it is recommended
+// to apply a large jitter to the first duration, and provide a small jitter for the
+// per-tick jitter.  This will ensure that operations started at similar times will
+// have varying initial interval states, while minimizing the amount of extra work
+// introduced by the per-tick jitter.
+type Interval struct {
+	cfg       IntervalConfig
+	ch        chan time.Time
+	closeOnce sync.Once
+	done      chan struct{}
+}
+
+// IntervalConfig configures an interval.  The only required parameter is
+// the Duration field which *must* be a positive duration.
+type IntervalConfig struct {
+	// Duration is the duration on which the interval "ticks" (if a jitter is
+	// applied, this represents the upper bound of the range).
+	Duration time.Duration
+
+	// FirstDuration is an optional special duration to be used for the first
+	// "tick" of the interval.  This duration is not jittered.
+	FirstDuration time.Duration
+
+	// Jitter is an optional jitter to be applied to each step of the interval.
+	// It is usually preferable to use NewSmallJitter for this parameter, since
+	// periodic operations are typically costly and the affect of the jitter is
+	// cumulative.
+	Jitter Jitter
+}
+
+// NewInterval creates a new interval instance.  This function panics on non-positive
+// interval durations (equivalent to time.NewTicker).
+func NewInterval(cfg IntervalConfig) *Interval {
+	if cfg.Duration <= 0 {
+		panic(errors.New("non-positive interval for NewInterval"))
+	}
+
+	ch := make(chan time.Time, 1)
+	done := make(chan struct{})
+
+	interval := &Interval{
+		ch:   ch,
+		cfg:  cfg,
+		done: done,
+	}
+
+	firstDuration := cfg.FirstDuration
+	if firstDuration == 0 {
+		firstDuration = interval.duration()
+	}
+
+	// start the timer in this goroutine to improve
+	// consistency of first tick.
+	timer := time.NewTimer(firstDuration)
+
+	go interval.run(timer)
+
+	return interval
+}
+
+// Stop permanently stops the interval.
+func (i *Interval) Stop() {
+	i.closeOnce.Do(func() {
+		close(i.done)
+	})
+}
+
+// Next is the channel over which the intervals are delivered.
+func (i *Interval) Next() <-chan time.Time {
+	return i.ch
+}
+
+// duration gets the duration of the interval.  Each call applies the jitter
+// if one was supplied.
+func (i *Interval) duration() time.Duration {
+	if i.cfg.Jitter == nil {
+		return i.cfg.Duration
+	}
+	return i.cfg.Jitter(i.cfg.Duration)
+}
+
+func (i *Interval) run(timer *time.Timer) {
+	defer timer.Stop()
```

Comment:
Not closing the output channel is deliberate and consistent with the behavior of `time.Ticker`.  From the docs of `time.Ticker`:

> Stop does not close the channel, to prevent a concurrent goroutine reading from the channel from seeing an erroneous "tick".

Reply:
Gotcha, can you leave a comment about that in `Interval.Stop` too?